services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -1279,12 +1279,13 @@ else if (context.getStatus() == RequiredActionContext.Status.CHALLENGE) {
             authSession.setAuthNote(AuthenticationProcessor.CURRENT_AUTHENTICATION_EXECUTION, model.getProviderId());
             return context.getChallenge();
         }
-        else if (context.getStatus() == RequiredActionContext.Status.SUCCESS) {
+        else if (context.getStatus() == RequiredActionContext.Status.SUCCESS || context.getStatus() == RequiredActionContext.Status.IGNORE) {
```

Comment:
Wouldn't be move the ignore in a different if and do the next only in that situation? Something like:

```diff
diff --git a/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java b/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
index 10fd0e7523..847341149f 100755
--- a/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
+++ b/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
@@ -1279,6 +1279,10 @@ public class AuthenticationManager {
             authSession.setAuthNote(AuthenticationProcessor.CURRENT_AUTHENTICATION_EXECUTION, model.getProviderId());
             return context.getChallenge();
         }
+        else if (context.getStatus() == RequiredActionContext.Status.IGNORE) {
+            authSession.getAuthenticatedUser().removeRequiredAction(factory.getId());
+            return nextActionAfterAuthentication(session, authSession, session.getContext().getConnection(), request, session.getContext().getUri(), event);
+        }
         else if (context.getStatus() == RequiredActionContext.Status.SUCCESS) {
             event.clone().event(EventType.CUSTOM_REQUIRED_ACTION).detail(Details.CUSTOM_REQUIRED_ACTION, factory.getId()).success();
             // don't have to perform the same action twice, so remove it from both the user and session required actions
```

Reply:
In the case where a user has already verified their email, and for some reason you add the `VERIFY_EMAIL` and `webauthn-register` (or any other action with low priority) action to the user, the `webauthn-register `action would be skipped because of `context.success()` [here](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/requiredactions/VerifyEmail.java#L61).

Reply:
Ah! Ok, the failure in fips was that the verify email was re-added all the time. OK. Thanks!

Reply:
Maybe then if ignored the event should not be added. I don't think we shoukd report an event if verify_email is ignored. The next line. WDYT?

Reply:
Thanks, ok about not reporting the event.

Reply:
I noticed two different behaviors based on the auth flow when the email verification is enabled on the realm and the email is allowed to be null by user profile:

With the browser flow, the verify_email action is first triggered but then skipped if the email is null, allowing the user to log in.

With direct grant, the verify_email action is first triggered and then the login is blocked because of this check:
https://github.com/graziang/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/grants/ResourceOwnerPasswordCredentialsGrantType.java#L118-L128

Reply:
we should consider if the current behavior of ignoring the VERIFY_EMAIL action in case of a null email is correct or if the user should be stopped with a message indicating that the email is not present and therefore not verified. WDYT?

Reply:
Yes, it seems that there is another test that fails now. As you say I think there is a discrepancy between `evaluateTriggers` and `requiredActionChallenge`. The former adds the action is no email but the realm is configured to verify them and the latter ignores it if there is no email. Maybe if the realm is set to verify email the action should fail if there is no email. So this part:

```java
        String email = context.getUser().getEmail();
        if (Validation.isBlank(email)) {
            context.ignore();
            return;
        }
```

Can be be changed to failure if the realm is configured to everybody to check the email and ignore if not. WDYT?

Reply:
I was thinking if it would be better to let it fail in any case if the email is null instead of ignoring it, because otherwise if you manually add the required action to the user the behavior between the the two flows would still be different

Reply:
It looks that is general fix for all required which call context.success(); from requiredActionChallenge(final RequiredActionContext context)

[context.success(); from requiredActionChallenge(final RequiredActionContext context)](https://github.com/keycloak/keycloak/issues/31161#issuecomment-2219769063)

@rmartinc Is it ok for you, are you going to approve this PR?

Reply:
I would do the same in both situations. The condition for `evaluateTriggers` is `context.getRealm().isVerifyEmail() && !context.getUser().isEmailVerified()` so I would do the same for `requiredActionChallenge`. The action should not pass if the email is blank but the realm is configured to verify email. I think it's the less disruptive decision. But, as you say, we can also decide to always fail if the action is assigned and the email is blank (but maybe you need to change tests).

Reply:
@radoslawsyga We need to take a decision about what to do with the verify email action. There is a test that is failing right now. It will be approved and merged when ready.

Reply:
Updated the PR to always fail if the action is assigned and the email is blank.  Tests look OK